File: ios/Classes/SRequest.m
//
//  SRequest.m
//  map4d_services
//
//  Created by Huy Dang on 1/20/22.
//

#import "SRequest.h"
#import "SConverter.h"

@implementation SRequest

#pragma mark - Geocode

+ (MFServiceRequest *)buildGeocodingRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/v2/geocode"];
  MFGeocodeParams *params = [[MFGeocodeParams alloc] initWithAddress:data[@"address"]];
  params.location = [SConverter toLocationComponent:data[@"location"]];
  params.viewbox = [SConverter toViewbox:data[@"viewbox"]];
  
  request.params = params;
  return request;
}

#pragma mark - Place

+ (MFServiceRequest *)buildPlaceDetailRequestWithId:(NSString *)placeId {
  NSString *path = [@"/sdk/place/detail/" stringByAppendingString:placeId];
  return [[MFServiceRequest alloc] initWithPath:path];
}

+ (MFServiceRequest *)buildTextSearchRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/place/text-search"];
  MFTextSearchParams *params = [[MFTextSearchParams alloc] initWithText:data[@"text"]];
  
  params.types = data[@"types"];
  params.location = [SConverter toLocationComponent:data[@"location"]];
  params.datetime = [SConverter toDate:data[@"datetime"]];
  
  request.params = params;
  return request;
}

+ (MFServiceRequest *)buildNearbySearchRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/place/nearby-search"];
  MFNearbySearchParams *params = [[MFNearbySearchParams alloc] initWithLocation:[SConverter toLocationComponent:data[@"location"]]
                                                                         radius:[SConverter toInt64:data[@"radius"]]
                                                                           text:data[@"text"]];
  params.types = data[@"types"];
  params.tags = data[@"tags"];
  params.datetime = [SConverter toDate:data[@"datetime"]];
  
  request.params = params;
  return request;
}

+ (MFServiceRequest *)buildViewboxSearchRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/place/viewbox-search"];
  MFViewboxSearchParams *params = [[MFViewboxSearchParams alloc] initWithViewbox:[SConverter toViewbox:data[@"viewbox"]]
                                                                            text:data[@"text"]];
  params.types = data[@"types"];
  params.tags = data[@"tags"];
  params.datetime = [SConverter toDate:data[@"datetime"]];
  
  request.params = params;
  return request;
}

+ (MFServiceRequest *)buildSuggestionsRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/autosuggest"];
  MFSuggestionParams *params = [[MFSuggestionParams alloc] initWithText:data[@"text"]];
  params.acronym = [SConverter toBool:data[@"acronym"]];
  params.location = [SConverter toLocationComponent:data[@"location"]];
  
  request.params = params;
  return request;
}

#pragma mark - Route

+ (MFServiceRequest *)buildDirectionsRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/route"];
  
  MFLocationComponent *origin = [SConverter toLocationComponent:data[@"origin"]];
  MFLocationComponent *destination = [SConverter toLocationComponent:data[@"destination"]];
  MFDirectionsParams *params = [[MFDirectionsParams alloc] initWithOrigin:origin destination:destination];
  params.waypoints = [SConverter toLocationComponentArray:data[@"points"]];
  params.mode = [SConverter toTravelMode:data[@"mode"]];
  params.language = [SConverter toLanguageResult:data[@"language"]];
  params.weighting = [SConverter toRouteWeighting:data[@"weighting"]];
  params.avoid = [SConverter toRouteAvoid:data[@"avoid"]];
  params.avoidRouteTypes = [SConverter toRouteTypeAvoidArray:data[@"avoidRoads"]];
  
  request.params = params;
  return request;
}

+ (MFServiceRequest *)buildRouteETARequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/route/eta"];
  
  NSArray<MFLocationComponent *> *origins = [SConverter toLocationComponentArray:data[@"origins"]];
  MFLocationComponent *destination = [SConverter toLocationComponent:data[@"destination"]];
  MFRouteETAParams *params = [[MFRouteETAParams alloc] initWithOrigins:origins destination:destination];
  params.mode = [SConverter toTravelMode:data[@"mode"]];
  params.language = [SConverter toLanguageResult:data[@"language"]];
  params.weighting = [SConverter toRouteWeighting:data[@"weighting"]];
  params.avoid = [SConverter toRouteAvoid:data[@"avoid"]];
  params.avoidRouteTypes = [SConverter toRouteTypeAvoidArray:data[@"avoidRoads"]];
  
  request.params = params;
  request.method = MFRequestMethodPost;
  return request;
}

+ (MFServiceRequest *)buildDistanceMatrixRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/route/matrix"];
  
  NSArray<MFLocationComponent *> *origins = [SConverter toLocationComponentArray:data[@"origins"]];
  NSArray<MFLocationComponent *> *destinations = [SConverter toLocationComponentArray:data[@"destinations"]];
  MFDistanceMatrixParams *params = [[MFDistanceMatrixParams alloc] initWithOrigins:origins destinations:destinations];
  params.mode = [SConverter toTravelMode:data[@"mode"]];
  params.language = [SConverter toLanguageResult:data[@"language"]];
  params.weighting = [SConverter toRouteWeighting:data[@"weighting"]];
  params.avoid = [SConverter toRouteAvoid:data[@"avoid"]];
  params.avoidRouteTypes = [SConverter toRouteTypeAvoidArray:data[@"avoidRoads"]];
  
  request.params = params;
  return request;
}

+ (MFServiceRequest *)buildGraphRouteRequestWithData:(NSDictionary *)data {
  MFServiceRequest *request = [[MFServiceRequest alloc] initWithPath:@"/sdk/route/graph"];
  
  NSArray<MFLocationComponent *> *points = [SConverter toLocationComponentArray:data[@"points"]];
  MFGraphRouteParams *params = [[MFGraphRouteParams alloc] initWithLocations:points];
  params.mode = [SConverter toTravelMode:data[@"mode"]];
  params.language = [SConverter toLanguageResult:data[@"language"]];
  params.weighting = [SConverter toRouteWeighting:data[@"weighting"]];
  params.avoid = [SConverter toRouteAvoid:data[@"avoid"]];
  params.avoidRouteTypes = [SConverter toRouteTypeAvoidArray:data[@"avoidRoads"]];
  
  request.params = params;
  return request;
}

@end

File: android/src/main/java/vn/map4d/services/map4d_services/SRequest.java
package vn.map4d.services.map4d_services;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import io.flutter.plugin.common.MethodCall;
import vn.map4d.services.MFMethod;
import vn.map4d.services.MFRequestParams;
import vn.map4d.services.MFServiceOptions;

class SRequest {
  static MFServiceOptions buildGeocodingRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/v2/geocode");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();
    String address = call.argument("address");
    if (address != null) {
      params.put("address", address);
    }

    String location = SConverter.toLocation(call.argument("location"));
    if (location != null) {
      params.put("location", location);
    }

    String viewbox = SConverter.toViewbox(call.argument("viewbox"));
    if (viewbox != null) {
      params.put("viewbox", viewbox);
    }
    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildPlaceDetailRequestWithId(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    StringBuilder urlBuilder = new StringBuilder("/sdk/place/detail");
    serviceOptionsBuilder.method(MFMethod.GET);
    String id = call.argument("id");
    if (id != null) {
      urlBuilder.append("/").append(id);
    }
    serviceOptionsBuilder.url(urlBuilder.toString());
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildTextSearchRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/place/text-search");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();
    String text = call.argument("text");
    if (text != null) {
      params.put("text", text);
    }

    String location = SConverter.toLocation(call.argument("location"));
    if (location != null) {
      params.put("location", location);
    }

    String types = SConverter.toTypes(call.argument("types"));
    if (types != null) {
      params.put("types", types);
    }

    String datetime = SConverter.toDatetime(call.argument("datetime"));
    if (datetime != null) {
      params.put("datetime", datetime);
    }
    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildNearbySearchRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/place/nearby-search");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();
    String location = SConverter.toLocation(call.argument("location"));
    if (location != null) {
      params.put("location", location);
    }

    String radius = SConverter.toRadius(call.argument("radius"));
    if (radius != null) {
      params.put("radius", radius);
    }

    String text = call.argument("text");
    if (text != null) {
      params.put("text", text);
    }

    String types = SConverter.toTypes(call.argument("types"));
    if (types != null) {
      params.put("types", types);
    }

    String tags = SConverter.toTags(call.argument("tags"));
    if (tags != null) {
      params.put("tags", tags);
    }

    String datetime = SConverter.toDatetime(call.argument("datetime"));
    if (datetime != null) {
      params.put("datetime", datetime);
    }
    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildViewboxSearchRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/place/viewbox-search");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();
    String viewbox = SConverter.toViewbox(call.argument("viewbox"));
    if (viewbox != null) {
      params.put("viewbox", viewbox);
    }

    String text = call.argument("text");
    if (text != null) {
      params.put("text", text);
    }

    String types = SConverter.toTypes(call.argument("types"));
    if (types != null) {
      params.put("types", types);
    }

    String tags = SConverter.toTags(call.argument("tags"));
    if (tags != null) {
      params.put("tags", tags);
    }

    String datetime = SConverter.toDatetime(call.argument("datetime"));
    if (datetime != null) {
      params.put("datetime", datetime);
    }
    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildSuggestionsRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/autosuggest");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();
    String text = call.argument("text");
    if (text != null) {
      params.put("text", text);
    }

    String acronym = SConverter.toAcronym(call.argument("acronym"));
    if (acronym != null) {
      params.put("acronym", acronym);
    }

    String location = SConverter.toLocation(call.argument("location"));
    if (location != null) {
      params.put("location", location);
    }
    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildDirectionsRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/route");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();

    String origin = SConverter.toLocation(call.argument("origin"));
    if (origin != null) {
      params.put("origin", origin);
    }

    String destination = SConverter.toLocation(call.argument("destination"));
    if (destination != null) {
      params.put("destination", destination);
    }

    String mode = SConverter.toMode(call.argument("mode"));
    if (mode != null) {
      params.put("mode", mode);
    }

    String weighting = SConverter.toWeighting(call.argument("weighting"));
    if (weighting != null) {
      params.put("weighting", weighting);
    }

    String language = SConverter.toLanguage(call.argument("language"));
    if (language != null) {
      params.put("language", language);
    }

    String points = SConverter.toLocationList(call.argument("points"));
    if (points != null) {
      params.put("points", points);
    }

    Object avoidObject = call.argument("restriction");
    String avoid = SConverter.toAvoid(avoidObject);
    if (avoid != null) {
      params.put("avoid", avoid);
    }

    String avoidRoads = SConverter.toAvoidRoads(avoidObject);
    if (avoidRoads != null) {
      params.put("avoidRoads", avoidRoads);
    }

    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildRouteETARequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/route/eta");
    serviceOptionsBuilder.method(MFMethod.POST);

    JSONObject jsonObject = new JSONObject();
    try {
      JSONArray origins = SConverter.toOriginJsonArray(call.argument("origins"));
      if (origins != null) {
        jsonObject.put("origins", origins);
      }

      String destination = SConverter.toLocation(call.argument("destination"));
      if (destination != null) {
        jsonObject.put("destination", destination);
      }

      String mode = SConverter.toMode(call.argument("mode"));
      if (mode != null) {
        jsonObject.put("mode", mode);
      }

      String weighting = SConverter.toWeighting(call.argument("weighting"));
      if (weighting != null) {
        jsonObject.put("weighting", weighting);
      }

      String language = SConverter.toLanguage(call.argument("language"));
      if (language != null) {
        jsonObject.put("language", language);
      }

      Object avoidObject = call.argument("restriction");
      String avoid = SConverter.toAvoid(avoidObject);
      if (avoid != null) {
        jsonObject.put("avoid", avoid);
      }

      String avoidRoads = SConverter.toAvoidRoads(avoidObject);
      if (avoidRoads != null) {
        jsonObject.put("avoidRoads", avoidRoads);
      }
    } catch (JSONException e) {
      e.printStackTrace();
    }

    serviceOptionsBuilder.json(jsonObject.toString());
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildDistanceMatrixRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/route/matrix");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();

    String origins = SConverter.toLocationList(call.argument("origins"));
    if (origins != null) {
      params.put("origins", origins);
    }

    String destinations = SConverter.toLocationList(call.argument("destinations"));
    if (destinations != null) {
      params.put("destinations", destinations);
    }

    String mode = SConverter.toMode(call.argument("mode"));
    if (mode != null) {
      params.put("mode", mode);
    }

    String weighting = SConverter.toWeighting(call.argument("weighting"));
    if (weighting != null) {
      params.put("weighting", weighting);
    }

    String language = SConverter.toLanguage(call.argument("language"));
    if (language != null) {
      params.put("language", language);
    }

    Object avoidObject = call.argument("restriction");
    String avoid = SConverter.toAvoid(avoidObject);
    if (avoid != null) {
      params.put("avoid", avoid);
    }

    String avoidRoads = SConverter.toAvoidRoads(avoidObject);
    if (avoidRoads != null) {
      params.put("avoidRoads", avoidRoads);
    }

    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }

  static MFServiceOptions buildGraphRouteRequestWithData(@NonNull MethodCall call) {
    MFServiceOptions.Builder serviceOptionsBuilder = new MFServiceOptions.Builder();
    serviceOptionsBuilder.url("/sdk/route/graph");
    serviceOptionsBuilder.method(MFMethod.GET);

    MFRequestParams params = new MFRequestParams();

    String points = SConverter.toLocationList(call.argument("points"));
    if (points != null) {
      params.put("points", points);
    }

    String mode = SConverter.toMode(call.argument("mode"));
    if (mode != null) {
      params.put("mode", mode);
    }

    String weighting = SConverter.toWeighting(call.argument("weighting"));
    if (weighting != null) {
      params.put("weighting", weighting);
    }

    String language = SConverter.toLanguage(call.argument("language"));
    if (language != null) {
      params.put("language", language);
    }

    Object avoidObject = call.argument("restriction");
    String avoid = SConverter.toAvoid(avoidObject);
    if (avoid != null) {
      params.put("avoid", avoid);
    }

    String avoidRoads = SConverter.toAvoidRoads(avoidObject);
    if (avoidRoads != null) {
      params.put("avoidRoads", avoidRoads);
    }

    serviceOptionsBuilder.requestParams(params);
    return serviceOptionsBuilder.build();
  }
}
